static-analysis-plugin/plugin/src/main/groovy/com/novoda/staticanalysis/CheckstyleConfigurator.groovy

```diff
@@ -0,0 +1,68 @@
+package com.novoda.staticanalysis
+
+import groovy.util.slurpersupport.GPathResult
+import org.gradle.api.GradleException
+import org.gradle.api.Project
+import org.gradle.api.plugins.quality.Checkstyle
+import org.gradle.internal.logging.ConsoleRenderer
+
+class CheckstyleConfigurator {
+
+    void configure(Project project, StaticAnalysisExtension extension) {
+        project.apply plugin: 'checkstyle'
+        project.afterEvaluate {
+            PenaltyExtension penalty = extension.penalty
+            boolean isAndroidApp = project.plugins.hasPlugin('com.android.application')
+            boolean isAndroidLib = project.plugins.hasPlugin('com.android.library')
+            if (isAndroidApp || isAndroidLib) {
+                def variants = isAndroidApp ? project.android.applicationVariants : project.android.libraryVariants
+                configureAndroid(project, variants)
+            }
+            Map<String, Integer> violations = [errors: 0, warnings: 0]
+            List<String> reports = []
+            project.tasks['check'].doLast {
+                if (violations.errors > penalty.maxErrors || violations.warnings > penalty.maxWarnings) {
+                    String message = "Checkstyle rule violations were found ($violations.errors errors, $violations.warnings warnings). " +
+                            "See the reports at: ${reports.collect { "\n* $it" }.join('')}"
+                    throw new GradleException(message)
+                }
+            }
+            project.tasks.withType(Checkstyle) { task ->
+                task.group = 'verification'
+                task.showViolations = false
+                task.ignoreFailures = true
+                task.doLast {
+                    File xmlReportFile = task.reports.xml.destination
+                    File htmlReportFile = new File(xmlReportFile.absolutePath - '.xml' + '.html')
+
+                    GPathResult xml = new XmlSlurper().parse(xmlReportFile)
+                    violations.errors += xml.'**'.findAll { node -> node.name() == 'error' && node.@severity == 'error' }.size()
```

Comment:
XPath (alike?) queries are used to scrape the xml report and collect all the violations for the final evaluation. This is one of the biggest changes in behaviour from the original `checkstyle` plugin: static analysis evaluation will be fully completed before the evaluation of the reports take place. This of course means that even if your build fails you still gonna have all the reports for all the source sets listing the violations encountered.